unittests/unittest_ip_resolver.cpp

```diff
@@ -0,0 +1,383 @@
+{
+    zmq::ip_addr_t addr;
+
+    memset (&addr, 0xba, sizeof (addr));
+
+    return addr;
+}
+
+static void validade_ipv6_addr (const zmq::ip_addr_t &addr_,
+                                const struct in6_addr &expected_addr_,
+                                uint16_t expected_port_ = 0,
+                                uint16_t expected_zone_ = 0)
+{
+    const sockaddr_in6 *ip6_addr = &addr_.ipv6;
+
+    TEST_ASSERT_EQUAL (AF_INET6, addr_.generic.sa_family);
+
+    int neq = memcmp (&ip6_addr->sin6_addr,
+                      &expected_addr_,
+                      sizeof (expected_addr_));
+
+    TEST_ASSERT_EQUAL (0, neq);
+    TEST_ASSERT_EQUAL (htons (expected_port_), ip6_addr->sin6_port);
+    TEST_ASSERT_EQUAL (expected_zone_, ip6_addr->sin6_scope_id);
+}
+
+static void validade_ipv4_addr (const zmq::ip_addr_t &addr_,
+                                const struct in_addr &expected_addr_,
+                                uint16_t expected_port_ = 0)
+{
+    const sockaddr_in *ip4_addr = &addr_.ipv4;
+
+    TEST_ASSERT_EQUAL (AF_INET, addr_.generic.sa_family);
+    TEST_ASSERT_EQUAL (expected_addr_.s_addr, ip4_addr->sin_addr.s_addr);
+    TEST_ASSERT_EQUAL (htons (expected_port_), ip4_addr->sin_port);
+}
+
+void test_bind_any (int ipv6_)
+{
+    zmq::ip_resolver_t resolver;
+    zmq::ip_addr_t addr = test_bad_addr ();
+    int rc;
+
+    resolver
+        .bindable (true)
+        .allow_dns (false)
+        .allow_nic_name (false)
+        .expect_port (true)
+        .ipv6 (ipv6_);
+
+    rc = resolver.resolve(&addr, "*:*");
+
+    TEST_ASSERT_EQUAL (0, rc);
+
+    if (ipv6_) {
+        struct in6_addr anyaddr;
+
+#ifdef ZMQ_HAVE_VXWORKS
+        anyaddr = IN6ADDR_ANY_INIT;
+#else
+        anyaddr = in6addr_any;
+#endif
+        validade_ipv6_addr (addr, anyaddr, 0);
+    } else {
+        struct in_addr anyaddr;
+
+        anyaddr.s_addr = htonl (INADDR_ANY);
+
+        validade_ipv4_addr (addr, anyaddr, 0);
+    }
+
+    //  Wildcard should be rejected if we're not looking for a
+    //  bindable address
+    resolver.bindable (false);
+
+    TEST_ASSERT_EQUAL(-1, resolver.resolve(&addr, "*:*"));
+    TEST_ASSERT_EQUAL(-1, resolver.resolve(&addr, "*:1234"));
+
+    // This however works. Should it ? For the time being I'm going to
+    // keep it that way for backcompat but I can't imagine why you'd
+    // want a wildcard port if you're not binding.
+    TEST_ASSERT_EQUAL(0, resolver.resolve(&addr, "127.0.0.1:*"));
+}
+
+void test_bind_any_ipv4 () {
+    test_bind_any (false);
+}
+
+void test_bind_any_ipv6 () {
+    test_bind_any (true);
+}
+
+void test_parse_ipv4 () {
+    zmq::ip_resolver_t resolver;
+    zmq::ip_addr_t addr = test_bad_addr ();
+    struct in_addr expected_addr;
+
+    resolver
+        .bindable (false)
+        .allow_dns (false)
+        .allow_nic_name (false)
+        .expect_port (false)
+        .ipv6 (false);
+
+    assert (inet_pton (AF_INET, "1.2.128.129", &expected_addr) == 1);
```

Comment:
this will break Windows XP, you'll want a function that does something like:

```
#if defined(ZMQ_HAVE_WINDOWS) && (_WIN32_WINNT < 0x0600)
    ip4addr.sin_addr.s_addr = inet_addr ("127.0.0.1");
#else
    inet_pton (AF_INET, "127.0.0.1", &ip4addr.sin_addr);
#endif
```